apollo-federation/src/sources/connect/validation/selection.rs

```diff
@@ -392,3 +394,105 @@ impl SelectionValidator<'_, '_> {
         self.path.last().unwrap_or(&self.root)
     }
 }
+
+/// When using `*`, it must be mapped to a field via an alias, and the field
+/// must be a non-list custom scalar.
+fn validate_star_selection<'schema>(
+    group: &Group<'schema>,
+    schema: &'schema Schema,
+    selection_coordinate: &str,
+    selection_location: &Option<Range<LineColumn>>,
+) -> Vec<Message> {
+    group
+        .selection
+        .star_iter()
+        .filter_map(|star| {
+            let Some(field_name) = star.alias().map(|a| a.name()) else {
+                return Some(Message {
+                    code: Code::InvalidStarSelection,
+                    message: format!(
+                        "{coordinate} contains `*` without an alias. Use `fieldName: *` to map properties to a field.",
+                        coordinate = selection_coordinate,
+                    ),
+                    locations: selection_location.iter().cloned().collect(),
+                });
+            };
+
+            let Some(definition) = group.ty.fields.get(field_name) else {
+                return Some(Message {
+                    code: Code::SelectedFieldNotFound,
+                    message: format!(
+                        "{coordinate} contains field `{field_name}`, which does not exist on `{parent_type}`.",
+                        coordinate = selection_coordinate,
+                        parent_type = group.ty.name,
+                    ),
+                    locations: selection_location.iter().cloned().collect(),
+                });
+            };
+
+            if definition.ty.is_list() {
+                return Some(Message {
+                    code: Code::InvalidStarSelection,
+                    message: format!(
+                        "{coordinate} contains `{field_name}: *` but the field `{parent_type}.{field_name}: {ty}` returns a list. It must be a non-list custom scalar.",
+                        coordinate = selection_coordinate,
+                        field_name = field_name,
+                        parent_type = group.ty.name,
+                        ty = definition.ty
+                    ),
+                    locations: selection_location.iter().cloned().collect(),
```

Comment:
I usually also include the location of the field definition, to point users at where they should modify the type.